### PR TITLE
Add Buildkite configuration

### PR DESCRIPTION
**Pull request by Mudit-G**

### CODE REVIEW

File: ci/build.sh [hunk 1]
```diff
@@ -4,12 +4,22 @@ set -e -u -x -o pipefail
...
   PLATFORM="windows"
+  BINARY="spatial.exe"
 fi
-spatial build --target $PLATFORM --log_level=debug
+
+
+curl -Ls -o "${BINARY}.tmp" "https://console.improbable.io/toolbelt/download/${SPATIALOS_TOOLBELT_VERSION}/${PLATFORM}" || exit 1
+chmod +x "${BINARY}.tmp"
+mv "${BINARY}.tmp" "${BINARY}"
+./$BINARY build --target $PLATFORM --log_level=debug
```
Comment (RNabel) [inline ci/build.sh]: `./` not needed.

File: .toolsharerc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pinned_tool:
+  - tool: "imp-ci"
+    version: "20200206.175502.ed41e47ac7"
+  - tool: "imp-tool"
+    version: "20200217.173900.11796cb28c"
```
Comment (RNabel) [inline .toolsharerc]: Nit: newline at end of file

File: ci/build.sh [hunk 3]
```diff
@@ -4,12 +4,55 @@ set -e -u -x -o pipefail
 
 cd "$(dirname "$0")/.."
 
+if [[ -z ${BUILDKITE+x} ]]; then
+  echo "This script is intended for internal Improbable use."
```
Comment (RNabel) [inline ci/build.sh]: Maybe: This script is intended to run in CI only?

File: ci/build.sh [hunk 4]
```diff
@@ -4,12 +4,55 @@ set -e -u -x -o pipefail
...
+  exit 0
```
Comment (RNabel) [inline ci/build.sh]: I'd argue that this is an error and should therefore return a non-zero exit code.

File: ci/build.sh [hunk 5]
```diff
@@ -4,12 +4,55 @@ set -e -u -x -o pipefail
...
+fi
+
+SPATIALOS_TOOLBELT_VERSION="latest"
 PLATFORM=""
+BINARY=""
```
Comment (RNabel) [inline ci/build.sh]: This variable should be called `SPATIAL_BINARY` or similar.

File: ci/build.sh (hunk 1 above)
Comment (RNabel) [reply ci/build.sh]: Ping :)

Comment (Mudit-G) [reply ci/build.sh]: It's needed to avoid having to add the directory to `PATH` on Windows.

File: .toolsharerc (hunk 2 above)
Comment (Mudit-G) [reply .toolsharerc]: Added, thanks.

File: ci/build.sh (hunk 3 above)
Comment (Mudit-G) [reply ci/build.sh]: Sure, changed.

File: ci/build.sh (hunk 4 above)
Comment (Mudit-G) [reply ci/build.sh]: Yep, changed.

File: ci/build.sh (hunk 5 above)
Comment (Mudit-G) [reply ci/build.sh]: Renamed to `SPATIAL_BINARY`

File: ci/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,20 @@
+FROM ubuntu:18.04
+
+RUN ["apt-get", "update"]
```
Comment (RNabel) [inline ci/Dockerfile]: Please merge all of these into one `RUN` and add `&& rm -rf /var/lib/apt/lists/*` which removes the temporary state of `apt` from the image.

File: ci/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+VOLUME /build/tools
+ENV TOOLS_DIR "/build/tools"
+
+VOLUME /build/auth
+ENV AUTH_DIR "/build/auth"
+
+# Add the spatial wrapper script to the PATH.
+ENV PATH "$PATH:/build/project/ci"
+
+WORKDIR /build/project
```
Comment (RNabel) [inline ci/Dockerfile]: Why are you setting `WORKDIR` twice?

File: ci/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,20 @@
+FROM ubuntu:18.04
+
+RUN ["apt-get", "update"]
+RUN ["apt-get", "install", "-y", "gcc"]
+RUN ["apt-get", "install", "-y", "g++"]
+RUN ["apt-get", "install", "-y", "cmake"]
+
+WORKDIR /build/
+COPY ./ ./project
```
Comment (RNabel) [inline ci/Dockerfile]: COPY and ADD should always be as far down in the Dockerfile as possible because they easily invalidate the following layers.

File: ci/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+VOLUME /build/tools
+ENV TOOLS_DIR "/build/tools"
+
+VOLUME /build/auth
+ENV AUTH_DIR "/build/auth"
+
+# Add the spatial wrapper script to the PATH.
```
Comment (RNabel) [inline ci/Dockerfile]: Where is spatial coming from? Since you're downloading it dynamically, you shouldn't couple the container to the download location of the `spatial` binary.

File: ci/build.sh (hunk 1 above)
Comment (Mudit-G) [reply ci/build.sh]: Changed to `spatial` which is a wrapper script I've put on the PATH (needed because we have CMake rules using `spatial`)

File: ci/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+RUN ["apt-get", "update"]
+RUN ["apt-get", "install", "-y", "gcc"]
+RUN ["apt-get", "install", "-y", "g++"]
+RUN ["apt-get", "install", "-y", "cmake"]
+
+WORKDIR /build/
+COPY ./ ./project
+
+VOLUME /build/tools
+ENV TOOLS_DIR "/build/tools"
```
Comment (RNabel) [inline ci/Dockerfile]: I would change this to force the user of the container to set the `TOOLS_DIR` environment variable using `--env` and not declare specific volumes within the Dockerfile.

At the usage-location, this will mean that the user has to specify:
```
--env TOOLS_DIR=/build/tools
--volume <something>:/build/tools
```

This will allow the user to change the bash script which relies on the presence of e.g. tools, but dynamically change where they are without rebaking the container.

File: ci/spatial.bat [hunk 11]
```diff
@@ -0,0 +1 @@
+bash "%~dp0spatial" %*
```
Comment (RNabel) [inline ci/spatial.bat]: Why is this needed?

File: ci/build.sh [hunk 12]
```diff
@@ -4,12 +4,83 @@ set -e -u -x -o pipefail
...
+    --gcs_credentials_type=auto \
+    --tools=imp-tool
+else
+  TOOLBELT_PLATFORM="win"
+  SPATIAL_PLATFORM="windows"
+  SPATIAL_BINARY="spatial.exe"
+fi
+
+# Create temporary directories for the spatial binary and authentication token.
+export TOOLS_DIR="$(pwd)/tools"
```
Comment (RNabel) [inline ci/build.sh]: Use `mktemp -d` instead of this + the trap below.

**Review by RNabel — COMMENTED** [review]
Nice! Few comments

File: ci/Dockerfile (hunk 9 above)
Comment (Mudit-G) [reply ci/Dockerfile]: This is the `spatial` wrapper script in the `ci` folder.

File: ci/spatial.bat (hunk 11 above)
Comment (Mudit-G) [reply ci/spatial.bat]: We have CMake rules that call `spatial` which means it must be on the PATH. On Windows the `ci/spatial` bash script isn't picked up as an executable. So we introduce this batch script which will be run on Windows when running `spatial`. It just forwards to the bash script.

File: ci/Dockerfile (hunk 6 above)
Comment (Mudit-G) [reply ci/Dockerfile]: Sure, done

File: ci/Dockerfile (hunk 8 above)
Comment (Mudit-G) [reply ci/Dockerfile]: Moved `Copy` to the bottom.

File: ci/Dockerfile (hunk 10 above)
Comment (Mudit-G) [reply ci/Dockerfile]: Makes sense, removed them.

File: ci/Dockerfile (hunk 7 above)
Comment (Mudit-G) [reply ci/Dockerfile]: Just for the copy command, removed it.

File: ci/build.sh (hunk 12 above)
Comment (Mudit-G) [reply ci/build.sh]: Done

File: ci/build.sh [hunk 13]
```diff
@@ -4,12 +4,85 @@ set -e -u -x -o pipefail
...
+    --gcs_credentials_type=auto \
+    --tools=imp-tool
+else
+  TOOLBELT_PLATFORM="win"
+  SPATIAL_PLATFORM="windows"
+  SPATIAL_BINARY="spatial.exe"
+fi
+
+# Create temporary directories for the spatial binary and authentication token.
+export TOOLS_DIR="$(mktemp -d)"
```
Comment (RNabel) [inline ci/build.sh]: Super nit: alpha sort these :P

File: ci/build.sh [hunk 14]
```diff
@@ -4,12 +4,85 @@ set -e -u -x -o pipefail
...
+  docker run \
+    --rm \
+    --volume "${TOOLS_DIR}":/build/tools \
+    --env TOOLS_DIR=/build/tools \
+    --volume "${AUTH_DIR}":/build/auth \
+    --env AUTH_DIR=/build/auth \
+    --volume $(realpath $(pwd)):/code \
+    --workdir /code \
+    c_exampe_project_image \
+    /bin/bash -c "export PATH=\"/code/ci:\${PATH}\"; \
```
Comment (RNabel) [inline ci/build.sh]: You can save the escapes if you use single-quote to surround the string:
```
'export PATH="/code/ci:${PATH}";\
spatial build --target '"$SPATIAL_PLATFORM"
```
Note the change from single to double quotes after `--target`. We can't quote the entire string in single-quotes since we still need `$SPATIAL_PLATFORM` to resolve.
https://stackoverflow.com/a/6697781

**Review by RNabel — APPROVED** [review]
Nice!

File: ci/build.sh (hunk 13 above)
Comment (Mudit-G) [reply ci/build.sh]: Yep, done :)

File: ci/build.sh (hunk 14 above)
Comment (Mudit-G) [reply ci/build.sh]: Nice, changed.
